scripts/install_plugins.sh

```diff
@@ -5,32 +5,84 @@ CURRENT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 source "$CURRENT_DIR/shared_functions.sh"
 
 clone() {
-	local plugin=$1
-	cd $SHARED_TPM_PATH &&
+	local plugin; plugin="$1"
+	cd "$SHARED_TPM_PATH" &&
 		GIT_TERMINAL_PROMPT=0 git clone --recursive $plugin
 }
 
+install_plugin_git() {
+	local plugin; plugin="$1"
+	local branch; branch=":${1##*:}"
+	plugin="${1%$branch}"
+	clone "$plugin" ||
+		 clone "git:$plugin"
+	if [ "${branch}" != ":${1}" ]; then #if exists branch/revision
+		(cd "$SHARED_TPM_PATH"/$(shared_plugin_name "$plugin") &&
+		git checkout -q "${branch#:}" 2>&1 >/dev/null)
+	fi
+}
+
 # tries cloning:
 # 1. plugin name directly - works if it's a valid git url
 # 2. expands the plugin name to point to a github repo and tries cloning again
-clone_plugin() {
-	local plugin=$1
+install_plugin_github() {
+	local plugin; plugin="$1"
+	local branch; branch=":${1##*:}"
+	plugin="${1%$branch}"
 	clone "$plugin" ||
 		clone "https://git::@github.com/$plugin"
+	if [ "${branch}" != ":${1}" ]; then #if exists branch/revision
+		(cd "$SHARED_TPM_PATH"/$(shared_plugin_name "$plugin") &&
+		git checkout -q "${branch#:}" 2>&1 >/dev/null)
+	fi
+}
+
+install_plugin_web() {
+	local plugin; plugin="$1"
+	local plugin_name; plugin_name="$(shared_plugin_name "$plugin")"
+	(cd "$SHARED_TPM_PATH" && mkdir "${plugin_name}" &&
+	cd "${plugin_name}" && (wget --no-check-certificate "${plugin}" ||
+	curl -k -s -O "${plugin}"))
+}
+
+install_plugin_local() {
+	local plugin; plugin="$1"
+	local plugin_name; plugin_name="$(shared_plugin_name "$plugin")"
+	(cd "$SHARED_TPM_PATH" && mkdir "${plugin_name}" &&
+	cd "${plugin_name}" && cp -r "${plugin}" .)
 }
 
 # pull new changes or clone plugin
 install_plugin() {
-	local plugin="$1"
-	local plugin_name="$(shared_plugin_name "$plugin")"
+	local plugin; plugin="$1"
+	local plugin_name; plugin_name="$(shared_plugin_name "$plugin")"
 
 	if plugin_already_installed "$plugin"; then
 		# plugin is already installed
 		echo_message "Already installed \"$plugin_name\""
 	else
-		# plugin wasn't cloned so far - clone it
 		echo_message "Installing \"$plugin_name\""
-		clone_plugin "$plugin" &&
+		local handler; handler="${1%%:*}:"
+		plugin="${1#$handler}"
+
+		case "${handler}" in
+			''|*/*) case "${1}" in
+					/*|~*|\$*)	install_plugin_local  "${1}" ;;
+					*)		install_plugin_github "${1}" ;;
+				esac ;;
+			gh*|github*|git@github.com*)	install_plugin_github "${plugin}" ;;
```

Comment:
It's still defaulting to github =), the additional handlers are optional and could be used to improve readability I'm a kind of dump people and forget everything all the time so I generally are grateful of additional hints even if they're repetitive. The code doesn't add an important overhead but if you guys find it annoying it can be removed =)